### PR TITLE
Feature - More Swift 3 Updates

### PR DESCRIPTION
**Pull request by cnoon**
This PR refactors some of the APIs to be more Swift 3 friendly. More details are broken out in the commit log. Let's chat through these changes in person to decide if we think these are all the best Swift 3 stylistic decisions.

### CODE REVIEW

File: Source/Parser.swift [hunk 1]
```diff
@@ -182,25 +149,56 @@ public class Parser {
...
+                } else {
+                    let error = error as NSError
+
+                    throw ParserError.deserialization(
+                        failureReason: "JSON data deserialization failed with error: \"\(error.description)\""
+                    )
+                }
+            }
+
+            return result
```
Comment (cnoon) [inline Source/Parser.swift]: This is a pretty big win here. We're able to eliminate one of the `parseProperties` APIs by combining them into one thus removing the possibility of ambiguity and allowing us to use `parseProperties(from:)`.

File: Source/ParserProperty.swift [hunk 2]
```diff
@@ -118,11 +118,11 @@ public class ParserPropertyMaker {
...
     */
     @discardableResult
-    public func propertyForKeyPath(
-        _ keyPath: String,
+    public func property(
+        forKeyPath keyPath: String,
         type: ParserPropertyType,
         optional: Bool = false,
-        decodedToType decodableType: Decodable.Type?)
+        decodableType: Decodable.Type?)
```
Comment (cnoon) [inline Source/ParserProperty.swift]: It seemed logical here to replace `decodedToType` with simply `decodableType`. At this point in the method signature, we've lost grammatical correctness. Whether we use `propertyForKeyPath` or `property(forKeyPath:)` is really the tough decision. Apple has contradictory info about this case in the API guidelines.

File: Tests/DecoderTests.swift [hunk 3]
```diff
@@ -33,11 +33,11 @@ class ValidDecoder: Decoder {
         self.toDictionary = toDictionary
     }
 
-    func decode(object: AnyObject) throws -> Any {
-        let result = try Parser.parseProperties(json: object) { make in
-            make.propertyForKeyPath("subUInt", type: .uInt)
-            make.propertyForKeyPath("subInt", type: .int)
-            make.propertyForKeyPath("subString", type: .string)
+    func decode(_ object: AnyObject) throws -> Any {
```
Comment (cnoon) [inline Tests/DecoderTests.swift]: IMO, the `object` portion of this API didn't provide any additional context and was not necessary. When looking through all the places where this API is actually called, having `object:` doesn't really tell you anything you don't already know.

File: Source/ParserProperty.swift [hunk 4]
```diff
@@ -27,19 +27,19 @@ import Foundation
...
+    - double:     Represents a Swift `Double` type.
+    - bool:       Represents a Swift `Bool` type.
+    - array:      Represents a Swift `Array` type.
+    - dictionary: Represents a Swift `Dictionary` type.
+    - url:        Represents a Swift `URL` type.
 */
 public enum ParserPropertyType {
     case string
-    case uInt
+    case uint
```
Comment (cnoon) [inline Source/ParserProperty.swift]: This seems more like the way you would write this to me. Could certainly be debated though.

File: Source/Parser.swift (hunk 1 above)
Comment (EricAppel) [inline Source/Parser.swift]: These were intentionally separate APIs in the past to add clarity and strong typing for the `Data` option being available. The design guidelines state that weak type information should be compensated for with a noun to describe its role. In this case we may consider splitting these back up into `parseProperties(fromJSON: ...` and `parseProperties(fromData: ...`. Thoughts?

File: Source/Parser.swift [hunk 5]
```diff
@@ -35,28 +35,28 @@ public class Parser {
     /**
         Parses json data at the specified key path into an object of type `T`. `T` must implement the `Decodable` protocol.
 
-        - parameter data:       An NSData object containing encoded json data.
-        - parameter forKeyPath: The json key path identifying the object to be parsed. Default is `""`.
+        - parameter atKeyPath:  The json key path identifying the object to be parsed. Default is `""`.
+        - parameter from:       A Data object containing encoded json data.
```
Comment (cnoon) [inline Source/Parser.swift]: Should be `keyPath` and `data` in the docs.

**Comment by cnoon** [comment]
I'm going to just close this PR. Instead we're going to move all these changes and others through in #10.

**Comment by acflavell** [comment]
Thanks Chris,


I was wondering what was happening (or not). Do you want to connect sometime next week?


Andrew


________________________________

**Pull request closed without merging**